src/parsers/texttracks/ttml/resolve_styles_inheritance.ts

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright 2015 CANAL+ Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import objectAssign from "object-assign";
+import log from "../../../log";
+import arrayFindIndex from "../../../utils/array_find_index";
+import arrayIncludes from "../../../utils/array_includes";
+import { IStyleObject } from "./get_styling";
+
+/**
+ * Transform all styles inheriting from other styles to the same styles but with
+ * the inheritance removed (by resolving those inheritance here).
+ *
+ * Note that the original style object is directly muted with every inheritance
+ * they had resolved and removed.
+ *
+ * To make a pseudo-code analogy this would be equivalent to transform those
+ * two classes:
+ * ```
+ * class A {
+ *   methodA() {}
+ * }
+ *
+ * class B extends A {
+ *   method B() {}
+ * }
+ * ```
+ * into the same two classes without inheritance:
+ * ```
+ * class A {
+ *   methodA() {}
+ * }
+ * class B {
+ *   methodA() {} // inherited from class A
+ *   methodB() {}
+ * }
+ * ```
+ *
+ * Doing this here allows to simplify further treatment of those styles.
+ * @param {Array.<Object>} styles
+ * @returns {Array.<Object>}
+ */
+export default function resolveStylesInheritance(
+  styles : IStyleObject[]
+) : void {
+  // keep track of all the indexes parsed to avoid infinite loops
+  const recursivelyBrowsedIndexes : number[] = [];
+  function resolveStyleInheritance(styleElt : IStyleObject, index : number) {
+    recursivelyBrowsedIndexes.push(index);
+    for (let j = 0; j < styleElt.extendsStyles.length; j++) {
+      const extendedStyleID = styleElt.extendsStyles[j];
+      const extendedStyleIndex = arrayFindIndex(styles,
+                                                (x) => x.id === extendedStyleID);
+      if (extendedStyleIndex < 0) {
+        log.warn("TTML Parser: unknown style inheritance: " + extendedStyleID);
+      } else {
+        const extendedStyle = styles[extendedStyleIndex];
+        if (arrayIncludes(recursivelyBrowsedIndexes, extendedStyleIndex)) {
```

Comment:
Si if i understand well, the extended style will not be treated in a new resolveStyleInheritence in that case, but will still get style from it ?

Reply:
If 1 inherit from 3, and 3 inherit from 1, 3 will get 1 styles (but without styles from 3) ?

Reply:
Let's dive into it.

If 1 inherit from 3, and 3 inherit from 1:
  1. we begin by looking into 1.
  2. we see that it inherits from 3
  3. we look into 3
  4. we see that it inherits from 1
  5. we do not look into 1's inheritance, as an infinite loop avoidance measure
  6. we still merge 1 into 3 (but keeping properties from 3 that are in both)
  7. no inheritance left for 3, we empty `extendsStyles` for it
  8. we come back to 1
  9. we merge 3 into 1  (but keeping properties from 3 that are in both)
10. no inheritance left for 1, we empty `extendsStyles` for it
11. we look into 2
12. [...] (resolving 2)
13. we look into 3
14. no inheritance left for it, as `extendsStyles` has already been emptied
15. finished!

So to resume, if 1 takes from 3 and 3 takes from 1, 3 will be merged into 1 and 1 will be merged into 3 which I think is what we would want.

I added an unit test specifically for that situation.

Reply:
Ok, clearer.